File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-5.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @param {number} k
     * @return {number}
     */
    kthSmallest(root, k) {
        const st = [];
        let cur = root;

        while(cur || st.length > 0) {
            while(cur) {
                st.push(cur);
                cur = cur.left;
            }

            const node = st.pop();

            if(--k == 0) return node.val;

            cur = node.right;
        }

        return -1;
    }
}

File: Data Structures & Algorithms/serialize-and-deserialize-binary-tree/submission-5.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Codec:
    
    # Encodes a tree to a single string.
    def serialize(self, root: Optional[TreeNode]) -> str:
        if not root:
            return ""
        
        res = []
        q = deque([root])

        while q:
            node = q.popleft()
            if node:
                res.append(str(node.val))
                q.append(node.left)
                q.append(node.right)
            else:
                res.append("#")
        
        return ",".join(res)
        
    # Decodes your encoded data to tree.
    def deserialize(self, data: str) -> Optional[TreeNode]:
        if not data:
            return None
        
        values = data.split(",")
        root = TreeNode(int(values[0]))
        q = deque([root])
        idx = 1

        while q:
            node = q.popleft()

            if values[idx] != "#":
                node.left = TreeNode(int(values[idx]))
                q.append(node.left)
            idx += 1

            if values[idx] != "#":
                node.right = TreeNode(int(values[idx]))
                q.append(node.right)
            idx += 1
        
        return root